ako-gateway-api/objects/gateway_store.go

```diff
@@ -604,36 +737,52 @@ func (g *GWLister) DeleteRouteServiceMappings(routeTypeNsName string) {
 	g.gwLock.Lock()
 	defer g.gwLock.Unlock()
 
-	_, svcNsNameList := g.GetRouteToService(routeTypeNsName)
-	for _, svcNsName := range svcNsNameList {
-		if found, routeTypeNsNameList := g.GetServiceToRoute(svcNsName); found {
-			routeTypeNsNameList = utils.Remove(routeTypeNsNameList, routeTypeNsName)
-			if len(routeTypeNsNameList) == 0 {
-				g.serviceToRoute.Delete(svcNsName)
-				continue
+	if found, svcNsNameList := g.routeToService.Get(routeTypeNsName); found {
+		for _, svcNsName := range svcNsNameList.([]string) {
+			if found, routeTypeNsNameList := g.serviceToRoute.Get(svcNsName); found {
+				routeTypeNsNameListObj := routeTypeNsNameList.([]string)
+				routeTypeNsNameListObj = utils.Remove(routeTypeNsNameListObj, routeTypeNsName)
+				if len(routeTypeNsNameListObj) == 0 {
+					g.serviceToRoute.Delete(svcNsName)
+				} else {
+					g.serviceToRoute.AddOrUpdate(svcNsName, routeTypeNsNameList)
+				}
+			}
+		}
+
+		if found, gwNsNameList := g.routeToGateway.Get(routeTypeNsName); found {
+			for _, gwNsName := range gwNsNameList.([]string) {
+				if found, gwSvcNsNameList := g.gatewayToService.Get(gwNsName); found {
+					gwSvcNsNameListObj := gwSvcNsNameList.([]string)
+					for _, svcNsName := range svcNsNameList.([]string) {
+						gwSvcNsNameListObj = utils.Remove(gwSvcNsNameListObj, svcNsName)
+					}
+					if len(gwSvcNsNameListObj) == 0 {
```

Comment:
Should we rename this function as we are deleting gatewaytoservice mappings as well ?

Reply:
this should be fine as services are related to gateways through routes always.